### PR TITLE
Handle S3 credential expiration more gracefully

### PR DESCRIPTION
**Pull request by jrbourbeau**
Closes https://github.com/nsidc/earthaccess/issues/353

### CODE REVIEW

File: earthaccess/store.py
Comment (jrbourbeau) [inline earthaccess/store.py:213]: Previously we were always returning the same `s3fs.S3FileSystem` instance, regardless of if the `provider=`, etc. has been changed. We now cache S3 credentials based on a unique `(concept_id, daac, provider, endpoint)` combination, which should be more correct.

File: tests/unit/test_store.py
Comment (jrbourbeau) [inline tests/unit/test_store.py:114]: The previous `ORNLCLOUD` provider should actually be `ORNL_CLOUD` but was covered up in the previous implementation `get_s3fs_session` implementation always returning the same `s3fs.S3FileSystem` instance.

**Comment by jrbourbeau** [comment]
@MattF-NSIDC @betolink I confirmed the changes here work as expected in practice (i.e. no longer getting the permissions errors reported in the https://github.com/nsidc/earthaccess/issues/353)

**Comment by MattF-NSIDC** [comment]
It may realistically be a while before I can review this! :cry:

File: earthaccess/store.py
Comment (mfisher87) [inline earthaccess/store.py:101]: I'd like to specify the key `Tuple` a little more strictly if possible. Is it always a 4-tuple? Might be worth having a type alias (`Location = Tuple[Something, SomethingElse, AnotherSomething, SomethingAgain]`). Or maybe better: an annotated namedtuple?

```python
class Location(typing.NamedTuple):
   concept_id: Something
   daac: SomethingAgain
   provider: AnotherSomething
   endpoint: SomethingElse
```

Opinion, not request: I feel giving names to the remaining types may also help with readability, i.e. what can we name `Dict[str, str]` to make it more clear what its purpose is?

File: earthaccess/store.py
Comment (mfisher87) [inline earthaccess/store.py:220]:
```suggestion
            # If cached credentials are expired, invalidate the cache
            delta = datetime.datetime.now() - dt_init
```

File: earthaccess/store.py
Comment (mfisher87) [inline earthaccess/store.py:216]:
```suggestion
            # Check for cached credentials
            dt_init, creds = self._s3_credentials[location]
```

File: earthaccess/store.py
Comment (mfisher87) [inline earthaccess/store.py:229]: Weak nit: I kind of prefer longer variable names like `s3_credentials` :) What was the rationale for the name change?

File: earthaccess/store.py
Comment (mfisher87) [reply earthaccess/store.py:213]: A brief comment about why all four of these things is needed would be useful here, I think. 

Naively, `concept_id`'s presence is a bit confusing to me since we're deriving the value of `provider` from it below (actually getting the credentials doesn't depend on `concept_id`). Could we reduce the size of the tuple (and elminate some of the if...elif chain below) if we did that derivation in advance if `provider is None`?

File: earthaccess/store.py
Comment (mfisher87) [inline earthaccess/store.py:229]: I really like the new `try`/`catch`/`else` alternative to this logic. Much more readable.

File: earthaccess/store.py
Comment (mfisher87) [inline earthaccess/store.py:236]: I'm curious... how was this previously passing type checking?

File: tests/unit/test_store.py
Comment (mfisher87) [inline tests/unit/test_store.py:103]: Nice! Do we still need to inherit from `unittest.TestCast` now that we're using assert statements?

**Review by mfisher87 — DISMISSED** [review]
Great work! I have a couple thoughts about readability, but no concerns about the behavior at the moment. 

I don't think the suggestions I made are significant enough to warrant a second review, so if you're ready to merge after going through them: :rocket:

File: earthaccess/store.py
Comment (mfisher87) [reply earthaccess/store.py:220]: Hmm. Is there a better way to write suggestions to avoid the false deletion? The raw suggestion markdown is:

``````
```suggestion
            # If cached credentials are expired, invalidate the cache
            delta = datetime.datetime.now() - dt_init
```
``````

**Comment by betolink** [comment]
related to this, just for reference, Alex Mandel from DevSeed mentioned this piece of code that handles more or less the same for the MAAP project https://github.com/MAAP-Project/gedi-subsetter/blob/58c87ab907bc3b121510eeabdeb2b8e63c7df999/src/gedi_subset/maapx.py#L55-L61

File: earthaccess/store.py
Comment (jrbourbeau) [reply earthaccess/store.py:216]: I think this context is already provided with the 

```python
# Get existing S3 credentials if we already have them
```

comment a few lines above

File: earthaccess/store.py
Comment (jrbourbeau) [reply earthaccess/store.py:229]: I also like more descriptive variable names when there's not much other context around, but when there is sufficient context, more succinct names tend to reduce the cognitive load on readers (at least in my personal opinion). 

In this case this method is all about S3, and this variable is coming from a method called `get_s3_credentials`, so my sense is there's enough context to know that `creds` are S3 credentials.

File: earthaccess/store.py
Comment (jrbourbeau) [reply earthaccess/store.py:236]: No idea 🤷 But I definitely see that we're returning `None` here but the type signature says this should be an `s3fs` file system

File: tests/unit/test_store.py
Comment (jrbourbeau) [reply tests/unit/test_store.py:103]: We're still using `unittest` style code elsewhere, for example `self.assertTrue(isinstance(store.auth, Auth))` a few line above

File: earthaccess/store.py
Comment (jrbourbeau) [reply earthaccess/store.py:101]: Given this is (at least currently) only used in one place, and connected to a private attribute, adding additional typing infrastructure like this seems like overkill to me. I'd prefer to wait until this is used in multiple places before adding extra objects that code readers need to know about/keep track of. (Maybe this point is loosely related to the `daac` vs `provider` conversation over in https://github.com/nsidc/earthaccess/issues/349)

**Comment by jrbourbeau** [comment]
> mentioned this piece of code that handles more or less the same for the MAAP project

Nice, thanks for pointing us to that @betolink. Good to know for future reference.

**Comment by jrbourbeau** [comment]
@mfisher87 I've addressed or responded to all your review comments (thank you for taking the time to review this PR). There are a few minor, subjective points where I think we have differing opinions / styles -- let me know if any of those are blocking comments.

File: earthaccess/store.py
Comment (MattF-NSIDC) [reply earthaccess/store.py:237]: Makes total sense to me, thanks for reasoning that out.

**Comment by MattF-NSIDC** [comment]
@jrbourbeau thanks! I don't feel any of those differing opinions / styles should be blockers, marked all as resolved! I think this is ready to go :100:

**Review by MattF-NSIDC — APPROVED** [review]

**Pull request merged**
